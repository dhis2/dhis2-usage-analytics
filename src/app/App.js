import React from 'react';
import DatePicker from 'material-ui/lib/date-picker/date-picker';
import RaisedButton from 'material-ui/lib/raised-button';
import DropDownMenu from 'material-ui/lib/DropDownMenu';
import MenuItem from 'material-ui/lib/menus/menu-item';
import ThemeManager from 'material-ui/lib/styles/theme-manager';
var injectTapEventPlugin = require("react-tap-event-plugin");
injectTapEventPlugin();

import HeaderBar from '../header-bar/HeaderBar.component';
import { config } from 'd2/lib/d2';
import Client from './Client.js';
import TopMenu from './Topmenu.js';
import MyRawTheme from '../colortheme';


const buttonstyle = {
    marginLeft: 10,
    borderBottom: 1,
    borderBottomStyle: 'solid',
    borderRight: 1,
    borderRightStyle: 'solid',
    borderColor: '#dadada',
    backgroundColor: '#FFFFFF',
    width: 190
};

let startDate = new Date();
startDate.setMonth(startDate.getMonth() - 4);
let endDate = new Date();
let startString = '';
let endString = '';
let update = true;
let interval = "WEEK";

export default React.createClass({
    propTypes: {
        name: React.PropTypes.string,
        d2: React.PropTypes.object
    },

    childContextTypes: {
        d2: React.PropTypes.object,
        muiTheme: React.PropTypes.object
    },

    getChildContext() {
        return {
            d2: this.props.d2,
            muiTheme: ThemeManager.getMuiTheme(MyRawTheme)
        };
    },

    getInitialState: function () {
        return this.state = {
            url: config.baseUrl + "/dataStatistics?startDate=" + this.formatDate(startDate) + "&endDate=" + this.formatDate(endDate) + "&interval=WEEK",
            interval: 'WEEK',
            category: 'Favorite views'
        };
    },
    updateButton: function () {
        var style = {
            marginTop: 15,
            marginRight: 10,
            marginLeft: 10,
            display: 'block',
            float: 'left'
        };
        return <div><RaisedButton label="Update" className="raised-button" primary={true} style={style}
                                  onClick={() => this.updateURL()}/></div>;
    },

    updateURL: function () {
        startString = this.formatDate(startDate);
        endString = this.formatDate(endDate);
        let tempUrl = config.baseUrl + "/dataStatistics?startDate=" + startString + "&endDate=" + endString + "&interval=" + interval/*this.state.interval*/;
        update = true;
        this.setState({url: tempUrl});
    },

    DatePicker: function () {
        var style = {
            label: {
                display: 'block',
                fontFamily: 'Roboto, sans-serif',
                fontSize: 19

            },
            marginTop: 10,
            marginLeft: 10,
            marginBottom: 20,
            datepicker: {
                textFieldStyle: 'yyyy-mm-dd'
            }

        };
        return ( <div style={style}>
            <label style={style.label} htmlFor="start"><b>Start date:</b></label>
            <DatePicker id="start" defaultDate={startDate} formatDate={this.formatDate} mode="landscape"
                        onChange={(event, value) => startDate = value}/>
            <label style={style.label} htmlFor="end"><b>End date:</b></label>
            <DatePicker id="end" defaultDate={endDate} formatDate={this.formatDate} mode="landscape"
                        onChange={(event, value) =>  endDate = value}/>
        </div>);
    },

    formatDate(date){

        return date.getFullYear() + "-" + (date.getMonth() + 1) + "-" + date.getDate();
    },

    dropdownIntervalMenu: function () {
        var style = {
            label: {
                display: 'block',
                fontFamily: 'Roboto, sans-serif',
                marginTop: 10,
                marginLeft: 10,
                fontSize: 19
            },
            menu: {
                width: 116
            }

        };
        return (<div>
            <label style={style.label} htmlFor="intervaldrop"><b>Interval:</b></label>
            <DropDownMenu id="intervaldrop" autoWidth={false} value={interval/*this.state.interval*/} style={buttonstyle}
                          onChange={(event, index, value) => interval = value/*this.setState({interval: value})*/}>
                <MenuItem value={'DAY'} primaryText="DAY"/>
                <MenuItem value={'WEEK'} primaryText="WEEK"/>
                <MenuItem value={'MONTH'} primaryText="MONTH"/>
                <MenuItem value={'YEAR'} primaryText="YEAR"/>
            </DropDownMenu></div>);

    },

    dropdownCategoryMenu: function () {
        var style = {
            label: {
                display: 'block',
                fontFamily: 'Roboto, sans-serif',
                marginTop: 10,
                marginLeft: 10,
                fontSize: 19,
            }, icon: {
                color: '#000000'
            }
        };
        return (<div>
            <label style={style.label} htmlFor="Categorydrop"><b>Category:</b></label>
            <DropDownMenu id="Categorydrop" autoWidth={false} style={buttonstyle} iconStyle={style.icon}
                          value={this.state.category}
                          onChange={(event, index, value) => this.setState({category: value})}>
                <MenuItem value={'Favorite views'} primaryText="FAVORITE VIEWS"/>
                <MenuItem value={'Favorite saved'} primaryText="FAVORITES"/>
                <MenuItem value={'Users'} primaryText="USERS"/>
                <MenuItem value={'Top favorites'} primaryText="TOP FAVORITES"/>
                <MenuItem value={'Data values'} primaryText="DATA VALUES"/>
            </DropDownMenu></div>);

    },

    title: function () {
        var style = {
            h1: {
                fontFamily: 'Roboto, sans-serif',
                textAlign: 'center',
                marginBottom: 30,
                fontSize: 21,
                fontWeight: "normal"
            }
        };
        return (
            <h1 style={style.h1}>Usage Analytics</h1>
        )
    },

    render()
    {
        var style = {
            sidebar: {
                backgroundColor: '#f3f3f3',
                overflowY: 'auto',
                overflowX: 'hidden',
                width: 256
            }
        };

        return (
            <div className="app-wrapper">
                <HeaderBar />
                {React.createElement(
                    'div',
                    {style: style.sidebar, className: 'left-bar'},
                    this.title(),
                    this.DatePicker(),
                    this.dropdownIntervalMenu(),
                    this.dropdownCategoryMenu(),
                    this.updateButton()
                )}
                <div className="main-content">
                    {this.state.category == 'Top favorites' ? <TopMenu /> : <Client source={this.state.url} category={this.state.category} updatepage={update}/>}
                    {update = false}
                </div>
            </div>
        );
    }
});