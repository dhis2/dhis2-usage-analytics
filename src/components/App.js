import React from 'react'
import { Provider } from 'react-redux'
import store from '../store'
import UsageAnalytics from './UsageAnalytics'
import '../index.css'

function App() {
    return (
        <Provider store={store}>
            <UsageAnalytics />
        </Provider>
    )
}

export default App
