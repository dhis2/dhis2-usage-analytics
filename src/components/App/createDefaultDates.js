import moment from 'moment'

/**
 * This creates the default start and end dates for the app state. The format
 * is YYYY-MM-DD because that's what a date input expects, and that's what these
 * dates will be used for.
 *
 * The default is to have the end date at the current time, and the start date
 * 4 months before that.
 */

const createDefaultDates = () => {
    const now = Date.now()
    const format = 'YYYY-MM-DD'

    const initialStartDate = moment(now)
        .subtract(4, 'months')
        .format(format)
    const initialEndDate = moment(now).format(format)

    return { initialStartDate, initialEndDate }
}

export default createDefaultDates
