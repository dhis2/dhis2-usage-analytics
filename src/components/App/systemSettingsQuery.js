const systemSettingsQuery = {
    systemSettings: {
        resource: 'systemSettings',
        params: {
            key: 'keyCountPassiveDashboardViewsInUsageAnalytics',
        },
    },
}

export default systemSettingsQuery
