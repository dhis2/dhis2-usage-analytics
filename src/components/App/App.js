import React, { useState } from 'react'
import { CssVariables, CssReset } from '@dhis2/ui'
import { FAVORITE_VIEWS } from '../../constants/categories.js'
import { WEEK } from '../../constants/intervals.js'
import { SUM } from '../../constants/aggregations.js'
import { ALL } from '../../constants/chartTypes.js'
import { CHART_VIEW } from '../../constants/eventTypes.js'
import { PS_25 } from '../../constants/pageSizes.js'
import { ASC } from '../../constants/sortOrders.js'
import FilterFields from '../FilterFields/index.js'
import {
    LayoutContainer,
    LayoutSidebar,
    LayoutContent,
} from '../Layout/index.js'
import CategoryField from '../CategoryField/index.js'
import AppTitle from '../AppTitle/index.js'
import Visualization from '../Visualization/index.js'
import './App.css'
import createDefaultDates from './createDefaultDates.js'

const App = () => {
    const { initialStartDate, initialEndDate } = createDefaultDates()

    // State
    const [aggregation, setAggregation] = useState(SUM)
    const [category, setCategory] = useState(FAVORITE_VIEWS)
    const [chartType, setChartType] = useState(ALL)
    const [endDate, setEndDate] = useState(initialEndDate)
    const [eventType, setEventType] = useState(CHART_VIEW)
    const [interval, setInterval] = useState(WEEK)
    const [pageSize, setPageSize] = useState(PS_25)
    const [sortOrder, setSortOrder] = useState(ASC)
    const [startDate, setStartDate] = useState(initialStartDate)

    /**
     * The rendering of the different date intervals in the charts and tables
     * will crash if the props and the data aren't in sync. To prevent stale
     * props from being used we're tracking the stale status of interval.
     */
    const [isIntervalStale, setIsIntervalStale] = useState(false)
    const setIsStaleAndInterval = interval => {
        setIsIntervalStale(true)
        setInterval(interval)
    }

    return (
        <React.Fragment>
            <CssVariables spacers />
            <CssReset />
            <LayoutContainer>
                <LayoutSidebar>
                    <AppTitle />
                    <CategoryField
                        category={category}
                        setCategory={setCategory}
                    />
                    <FilterFields
                        aggregation={aggregation}
                        category={category}
                        chartType={chartType}
                        endDate={endDate}
                        eventType={eventType}
                        interval={interval}
                        pageSize={pageSize}
                        setAggregation={setAggregation}
                        setChartType={setChartType}
                        setEndDate={setEndDate}
                        setEventType={setEventType}
                        setInterval={setIsStaleAndInterval}
                        setPageSize={setPageSize}
                        setSortOrder={setSortOrder}
                        setStartDate={setStartDate}
                        sortOrder={sortOrder}
                        startDate={startDate}
                    />
                </LayoutSidebar>
                <LayoutContent>
                    <Visualization
                        aggregation={aggregation}
                        category={category}
                        chartType={chartType}
                        endDate={endDate}
                        eventType={eventType}
                        interval={interval}
                        isIntervalStale={isIntervalStale}
                        pageSize={pageSize}
                        setIsIntervalStale={setIsIntervalStale}
                        sortOrder={sortOrder}
                        startDate={startDate}
                    />
                </LayoutContent>
            </LayoutContainer>
        </React.Fragment>
    )
}

export default App
