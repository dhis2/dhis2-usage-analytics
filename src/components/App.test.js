import React from 'react'
import ReactDOM from 'react-dom'
import { shallow } from 'enzyme'
import App from './App'

jest.mock('react-redux', () => ({
    connect: jest.fn(() => component => `Connected${component.name}`),
    Provider: () => <div />, // eslint-disable-line react/display-name
}))

const mockFetch = jest.fn(async () => ({
    ok: true,
    json: async () => ({}),
}))

beforeEach(() => {
    jest.resetModules()
})

describe('<App/>', () => {
    it('renders without crashing', () => {
        window.fetch = mockFetch
        const div = document.createElement('div')

        ReactDOM.render(<App />, div)
        ReactDOM.unmountComponentAtNode(div)
    })

    it('renders the expected tree', () => {
        const tree = shallow(<App />)

        expect(tree).toMatchSnapshot()
    })
})
