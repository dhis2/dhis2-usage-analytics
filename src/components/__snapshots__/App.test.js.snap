// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> Matches the snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <UI
    theme="blue"
  >
    <HeaderBarContainer
      appName="Usage Analytics"
    />
    <Connect(UsageAnalytics) />
  </UI>
</Provider>
`;
